zcash_primitives/src/transaction/components/amount.rs

```diff
@@ -0,0 +1,179 @@
+use std::iter::Sum;
+use std::ops::{Add, AddAssign, Sub, SubAssign};
+
+const COIN: i64 = 1_0000_0000;
+const MAX_MONEY: i64 = 21_000_000 * COIN;
+
+pub const DEFAULT_FEE: Amount = Amount(10000);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Amount(i64);
```

Comment:
It would be nice if this had a doc-comment explaining the purpose of the newtype and explicitly declaring any invariants it maintains.  For instance in `from_i64` below, the constructor enforces some requirements on the input --  are those invariants that the type maintains, or just a requirement for that constructor?

Reply:
The main reason for the newtype is type safety. The requirements in the constructor are somewhat tricky however. `MAX_MONEY` is not actually the maximum amount of money in the consensus rules due to rounding errors (and in fact that will decrease slightly with the Blossom NU). It _is_ a consensus-critical validity check on visible values (transparent outputs, `vpub_old`, `vpub_new`, and `valueBalance`).

Now, we could have `Amount` maintain `{-MAX_MONEY..MAX_MONEY}` as an invariant, but what would this imply for the API? I'm not keen on burdening the API with error checks for every addition, given that the network consensus rules mean that any transaction created leveraging out-of-range values will be unmineable. But is a `panic!()` a reasonable way to handle this case? Wrapping definitely isn't (and is partly the motivation for checking `MAX_MONEY` in the constructor, so we know that any values instantiated are far below the `u64` upper end, and thus can assume there will be no overflow during transaction building).

I'll add a doc-comment in any case.

Reply:
I think we should panic if addition or subtraction overflows the range `-MAX_MONEY..MAX_MONEY`.